### PR TITLE
MAST: Catalog and Metadata Query Additions

### PR DESCRIPTION
**Pull request by phicc**
### Overview
Addition of Metadata query and catalogs.MAST PanSTARRS query methods along with associated documentation and unit tests

### Contributors
@phicc, @ceb8 

### Implementation
- Metadata query
  - `get_metadata` function in ObservationsClass and associated docs and tests
    - As part of this added `_get_colconfig_metadata` and  `_parse_type` functions
- Catalog.MAST PanSTARRS
  - Addition of PanSTARRS catalog option to catalog queries, including associated documentation and unit tests
  - New methods and variables:
    - `self._MAST_CATALOGS_REQUEST_URL` and ` self._MAST_CATALOGS_SERVICES` - catalogs.MAST URL information and string composition for URL (with required arguments)
    - `_fabric_json_to_table` - translate catalogs.MAST "fabric" microservice output to a valid Table
    - `_fabric_request` - simplified `_request` wrapper to be used for "fabric" microservice queries
    - `catalogs_service_request_async` - request method for catalogs.MAST catalogs (PanSTARRS) accepting the service (currently only supporting PanSTARRS), parameter criteria, page size and page. Returning a list of responses, similar to current functionality
    - `_build_catalogs_params` - composes the request `POST` body for catalog.MAST "fabric" microservice requests returning an array of tuples representing the parameters (parameters may exist in multiples i.e. `columns`)
    - `_check_catalogs_criteria_params` - confirms that at least one passed parameter is a criteria (non query modifiers) parameter

### General
  - Stylistic fixes in Iva's cloud access documentation
  - Fixed failing tests by increasing radius of search
  - Fixed failing `get_session` test that had a bad argument
  - Added continuation dots to python examples in docs
  - New `_generate_uuid_string` used for cache breaking Mashup services

### CODE REVIEW

**Comment by bsipocz** [comment]
@phicc - as I see the latest travis runs are failing due to PEP8 issues. These are listed at the end of the test run: https://travis-ci.org/astropy/astroquery/jobs/536570350#L1375
A useful workaround is to set up a flake8 like plugin in your editor either to auto fix the trivial issues or at least make a notification, most of them have a solution for it.

**Comment by phicc** [comment]
Thanks @bsipocz , didn't realize additional PEP8 items would be listed there. Updated and increased unit tests coverage.

**Comment by ceb8** [comment]
@bsipocz There's a mismatch between the milestone label (0.4.0) and what's in the changelog (0.3.10), but I'm not sure which one's the correct one. Can you clarify? Thanks.

**Comment by bsipocz** [comment]
@ceb8 - ignore the changelog for now. I was hoping to get 0.3.10 out of the door yesterday, so started to remilestone PRs that didn't seem to make it to that one. At the end there was a blocker that I couldn't yet resolve, so we ended up in this no men's land between the two releases.

**Comment by ceb8** [comment]
@bsipocz  OK, sounds good. So I think this PR is passing and ready for review then.

**Comment by bsipocz** [comment]
Thanks @ceb8! Adam or I will give a quick skim through then before merging.

File: astroquery/mast/core.py [hunk 1]
```diff
@@ -261,6 +394,50 @@ def session_info(self, silent=None, verbose=None):
...
+        Returns
+        -------
+        response : `~requests.Response`
+            The response from the server.
+        """
+
+        start_time = time.time()
+
+        response = super(MastClass, self)._request(method, url, params=params, data=data, headers=headers,
+                                                   files=files, cache=cache, stream=stream, auth=auth)
```
Comment (keflavich) [inline astroquery/mast/core.py]: why is this using the parent class method?  Is there something about the internal method we don't want to use?

File: astroquery/mast/core.py [hunk 2]
```diff
@@ -500,6 +687,150 @@ def service_request_async(self, service, params, pagesize=None, page=None, **kwa
 
         return response
 
+    @class_or_instance
+    def catalogs_service_request_async(self, service, params, page_size=None, page=None, **kwargs):
+        """
+           Given a MAST fabric service and parameters, builds and excecutes a fabric microservice catalog query.
```
Comment (keflavich) [inline astroquery/mast/core.py]: the docstring should be indented the same amount as the code (note that the closing """s are indented below)

File: docs/mast/mast.rst [hunk 3]
```diff
@@ -477,7 +546,40 @@ The TESS Input Catalog (TIC, and Disk Detective Catalog can also be queried base
...
+
+Columns returned from the query may be submitted with the columns parameter as a list of column names.
+
+The query may be sorted  with the sort_by parameter composed of either a single column name (to sort ascending),
+or a list of multiple column names and/or tuples of direction and column name (ASC/DESC, column name).
+
+To filter the query, criteria per column name are accepted. The 'AND' operation is performed between all
+column name criteria, and the 'OR' operation is performed within column name criteria. Per each column name
+parameter, criteria may consist of either a value or a list. The list may consist of a mix of values and
+tuples of criteria decorator (min, gte, gt, max, lte, lt, like, contains) and value,
```
Comment (keflavich) [inline docs/mast/mast.rst]: this line ends with a comma.  Could you add an example of a list of values + criteria tuples?

File: astroquery/mast/core.py (hunk 1 above)
Comment (phicc) [reply astroquery/mast/core.py]: We're not using the parent method here as we are in the process of refactoring our current use of `_request`. In the future we plan on renaming the current class `_request` to `_mashup_request` to keep mashup specific request items separate and allowing future API (most likely to MAST fabric services) methods to use the simpler `_fabric_request`

File: docs/mast/mast.rst (hunk 3 above)
Comment (phicc) [reply docs/mast/mast.rst]: Updated to end on a period as well as updated the example for the simple use of a criteria decorator.

File: astroquery/mast/core.py (hunk 2 above)
Comment (phicc) [reply astroquery/mast/core.py]: Updated indentation

**Comment by phicc** [comment]
@keflavich Would not making the requested change for the `_fabric_request` method be alright for this pull request as per my comment above? I have also updated the necessary documentation.

**Comment by bsipocz** [comment]
Thank you @phicc and @ceb8! and @keflavich for the review.

**Pull request merged**